### DISCUSSION TITLE
default logging macro level does not "propagate"?

### DISCUSSION THREAD

**Discussion by broland29** [gabime/spdlog#3031] [Q&A] [answered]
Hello. I am trying to use `spdlog` for my C++ project for the first time. It is hard to explain what my issue is, so here is a minimal example. I set the default logger level in `Logger.h` but the macro in `Other.cpp` does not print, even though in `Main.cpp` it prints. Why is this? How could I get the other macro print as well?

Logger.h
```
#pragma once

#define SPDLOG_ACTIVE_LEVEL SPDLOG_LEVEL_TRACE  // needed for macro usage, before spdlog.h!
#include <spdlog/spdlog.h>
#include "spdlog/sinks/stdout_color_sinks.h"


class Logger
{
public:
    static void initialize();
};
```
Logger.cpp
```
#include "Logger.h"


void Logger::initialize()
{
    auto logger = spdlog::stdout_color_mt("opencb_logger");

    logger->set_pattern("[%^%L%$] [%s:%#] [%!()] %v");
    spdlog::set_level(spdlog::level::trace);
    spdlog::set_default_logger(logger);

    SPDLOG_INFO("Logger initialized.");
}
```
Other.h:
```
#pragma once

#include <spdlog/spdlog.h>

class Other
{
public:
	Other();
};
```
Other.cpp
```
#include "Other.h"

Other::Other()
{
	SPDLOG_TRACE("other: SPDLOG_TRACE");
	spdlog::trace("other: spdlog::trace");
	SPDLOG_INFO("other: SPDLOG_INFO");
	spdlog::info("other: spdlog::info");
}
```

Main.cpp
```
#include "Logger.h"
#include "Other.h"

int main()
{
	Logger::initialize();
	SPDLOG_TRACE("main: SPDLOG_TRACE");
	spdlog::trace("main: spdlog::trace");
	SPDLOG_INFO("main: SPDLOG_INFO");
	spdlog::info("main: spdlog::info");
	Other other;
}
```

Output:
```
[I] [Logger.cpp:12] [Logger::initialize()] Logger initialized.
[T] [Main.cpp:7] [main()] main: SPDLOG_TRACE
[T] [:] [()] main: spdlog::trace
[I] [Main.cpp:9] [main()] main: SPDLOG_INFO
[I] [:] [()] main: spdlog::info
[T] [:] [()] other: spdlog::trace
[I] [Other.cpp:7] [Other::Other()] other: SPDLOG_INFO
[I] [:] [()] other: spdlog::info
```
I want to see line a line similar `[T] [Other.cpp:5] [Other::Other()] other: SPDLOG_TRACE` as well.

Thanks :smile:

**Comment by tt4g** [comment, marked as answer]
`#define SPDLOG_ACTIVE_LEVEL SPDLOG_LEVEL_TRACE` is not evaluated because there is no `#include "Logger.h"` in both `Other.h` and `Other.cpp`.
`SPDLOG_ACTIVE_LEVEL` is `SPDLOG_LEVEL_INFO` at the time `Other` is compiled.

You should define `SPDLOG_ACTIVE_LEVEL` in a header file that is `#include` in all source files, or define by compile option.

**Reply by broland29** [reply]
Thank you very much! I added the line `add_compile_definitions(SPDLOG_ACTIVE_LEVEL=SPDLOG_LEVEL_TRACE)` to my `CMakeLists.txt` and now it works as intended!
